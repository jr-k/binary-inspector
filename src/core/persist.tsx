const queryString = require("query-string");

const data = queryString.parse(window.location.hash);

export function set(key: string, value: string) {
  data[key] = value;
  window.location.hash = queryString.stringify(data);
}

export function get(key: string): undefined | string {
  return data[key];
}

export function has(key: string): boolean {
  return data[key] ? true : false;
}
