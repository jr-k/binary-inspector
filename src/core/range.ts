import { assert } from "chai";

/**
 * This represents a byte range in an ArrayBuffer. This can be used to read out
 * data as well as for highlighting regions in the BinaryView.
 */
export class ByteRange {
  buffer: ArrayBuffer;
  byteStart: number;
  byteLength: number;

  constructor(buffer: ArrayBuffer, byteStart: number, byteLength: number) {
    this.buffer = buffer;
    this.byteStart = byteStart;
    this.byteLength = byteLength;
  }

  toDataView() {
    return new DataView(this.buffer, this.byteStart, this.byteLength);
  }

  toUint8Array() {
    return new Uint8Array(this.buffer, this.byteStart, this.byteLength);
  }

  toHex(): string {
    return Array.from(this.toUint8Array())
      .map(b => b.toString(16).padStart(2, "0"))
      .join("");
  }

  bytes(byteStart: number, byteLength?: number) {
    if (byteLength == null) {
      byteLength = this.byteLength - byteStart;
    }

    if (byteStart < 0)
      throw new Error(`byteStart is ${byteStart}, but should be non-negative.`);
    if (byteStart + byteLength > this.byteLength)
      throw new Error(`New subrange does not fit within current subrange.`);

    return new ByteRange(this.buffer, this.byteStart + byteStart, byteLength);
  }

  bits(bitStart: number, bits: number) {
    return new BitRange(this.buffer, this.byteStart * 8 + bitStart, bits);
  }

  readBits(): Array<boolean> {
    let bits: Array<boolean> = [];
    let bytes = this.toUint8Array();
    while (bits.length < this.byteLength * 8) {
      let byte = bytes[Math.floor(bits.length / 8)];
      let pos = bits.length % 8;

      bits.push(((byte << pos) & 128) === 128);
    }
    return bits;
  }

  chunks(size: number): Array<ByteRange> {
    let chunks: Array<ByteRange> = [];
    let cursor = 0;
    while (cursor < this.byteLength) {
      chunks.push(this.bytes(cursor, Math.min(size, this.byteLength - cursor)));
      cursor += size;
    }
    return chunks;
  }
}

/**
 * This represents a range of bits in an ArrayBuffer. Use for reading bitfields.
 */
export class BitRange {
  buffer: ArrayBuffer;
  bitStart: number;
  bits: number;

  constructor(buffer: ArrayBuffer, bitStart: number, bits: number) {
    this.buffer = buffer;
    this.bitStart = bitStart;
    this.bits = bits;
  }

  enclosingByteRange(): ByteRange {
    return new ByteRange(
      this.buffer,
      Math.floor(this.bitStart / 8),
      Math.floor((this.bitStart + this.bits - 1) / 8) + 1
    );
  }

  readBits(): Array<boolean> {
    let bits = this.enclosingByteRange().readBits();
    return bits.slice(this.bitStart, this.bitStart + this.bits);
  }

  readUIntBE(): number {
    let number = 0;
    let bits = this.readBits();
    for (let i = 0; i < bits.length; ++i) {
      if (bits[bits.length - 1 - i]) number += Math.pow(2, i);
    }
    return number;
  }

  readBool(): boolean {
    let bits = this.readBits();
    assert.equal(bits.length, 1);
    return bits[0];
  }
}
