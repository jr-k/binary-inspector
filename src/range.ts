import {assert} from 'chai';

export class ByteRange {
  buffer: ArrayBuffer;
  byteStart: number;
  byteLength: number;

  constructor(buffer: ArrayBuffer, byteStart: number, byteLength: number) {
    this.buffer = buffer;
    this.byteStart = byteStart;
    this.byteLength = byteLength;
  }

  toDataView() {
    return new DataView(this.buffer, this.byteStart, this.byteLength);
  }

  toUint8Array() {
    return new Uint8Array(this.buffer, this.byteStart, this.byteLength);
  }

  toHex(): string {
    return Array.from(this.toUint8Array()).map (b => b.toString (16).padStart (2, "0")).join("");
  }

  bytes(byteStart: number, byteLength: number) {
    return new ByteRange(this.buffer, this.byteStart + byteStart, byteLength);
  }

  bits(bitStart: number, bits: number) {
    return new BitRange(this.buffer, this.byteStart * 8 + bitStart, bits);
  }

  readBits(): Array<boolean> {
    let bits = [];
    let bytes = this.toUint8Array();
    while (bits.length < this.byteLength * 8) {
      let byte = bytes[Math.floor(bits.length / 8)];
      let pos = bits.length % 8;

      bits.push(((byte << pos) & 128) === 128);
    }
    return bits;
  }
}

export class BitRange {
  buffer: ArrayBuffer;
  bitStart: number;
  bits: number;

  constructor(buffer: ArrayBuffer, bitStart: number, bits: number) {
    this.buffer = buffer;
    this.bitStart = bitStart;
    this.bits = bits;
  }

  enclosingByteRange(): ByteRange {
    return new ByteRange(this.buffer, Math.floor(this.bitStart / 8), Math.floor((this.bitStart + this.bits - 1) / 8) + 1);
  }

  readBits(): Array<boolean> {
    let bits = this.enclosingByteRange().readBits();
    return bits.slice(this.bitStart, this.bitStart + this.bits);
  }

  readUIntBE(): number {
    let number = 0;
    let bits = this.readBits();
    for(let i = 0; i < bits.length; ++i) {
      if (bits[bits.length - 1 - i])
        number += Math.pow(2, i)
    }
    return number;
  }

  readBool(): boolean {
    let bits = this.readBits();
    assert.equal(bits.length, 1);
    return bits[0];
  }
}
