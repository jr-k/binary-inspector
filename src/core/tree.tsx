import { ByteRange, BitRange } from "./range";

export class Tree {
  label: string;
  range: ByteRange | BitRange;
  children: Array<Tree>;

  constructor(
    label: string,
    range: ByteRange | BitRange,
    children: Array<Tree> = []
  ) {
    this.label = label;
    this.range = range;
    this.children = children;
  }
}
