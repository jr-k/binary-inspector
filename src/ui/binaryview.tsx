import * as React from "react";
import { Col, Form, Row } from "react-bootstrap";

import { ByteRange, BitRange } from "../core/range";

const HEX_BYTES_PER_ROW = 16;
const HEX_BYTES_PER_GROUP = 8;

const BINARY_BYTES_PER_ROW = 8;
const BINARY_BYTES_PER_GROUP = 4;

const DEFAULT_MAX_ROWS = 1000;

const SELECTED_COLOR = "#dedede";

type Format = "hex" | "binary";

export class BinaryView extends React.Component<
  { data: ByteRange; selected?: ByteRange | BitRange; maxRows: number },
  { format: Format }
> {
  static defaultProps = {
    maxRows: DEFAULT_MAX_ROWS
  };

  constructor(props) {
    super(props);
    this.state = { format: "hex" };
  }

  setFormat(format: Format) {
    this.setState({ format: format });
  }

  render() {
    return (
      <>
        <Form>
          <Form.Group as={Row}>
            <Form.Label column sm={2}>
              Format
            </Form.Label>
            <Col sm={10}>
              <Form.Control
                as="select"
                value={this.state.format}
                onChange={e => {
                  // @ts-ignore
                  this.setFormat(e.target.value);
                }}
              >
                <option value="hex">Hex</option>
                <option value="binary">Binary</option>
              </Form.Control>
            </Col>
          </Form.Group>
        </Form>
        {this.state.format === "hex" && this.renderHex()}
        {this.state.format === "binary" && this.renderBinary()}
      </>
    );
  }

  renderHex() {
    let data = this.props.data;

    // The total number of rows required to render the data.
    let totalRows = Math.ceil(data.byteLength / HEX_BYTES_PER_ROW);

    // The rows that we are rendering. If the data is small, this is all the rows.
    let rowStart = 0;
    let rowEnd = Math.ceil(data.byteLength / HEX_BYTES_PER_ROW);

    if (totalRows > this.props.maxRows) {
      if (this.props.selected) {
        const byteStart =
          this.props.selected instanceof ByteRange
            ? this.props.selected.byteStart
            : this.props.selected.enclosingByteRange().byteStart;
        const selectionRow = Math.floor(byteStart / HEX_BYTES_PER_ROW);

        rowStart = Math.max(
          0,
          selectionRow - Math.floor((this.props.maxRows - 1) / 2)
        );
        rowEnd = rowStart + this.props.maxRows;
      } else {
        rowStart = 0;
        rowEnd = rowStart + this.props.maxRows;
      }
    }

    const bytesLeft = data.byteLength - rowStart * HEX_BYTES_PER_ROW;
    data = data.bytes(
      rowStart * HEX_BYTES_PER_ROW,
      Math.min(bytesLeft, (rowEnd - rowStart) * HEX_BYTES_PER_ROW)
    );

    return (
      <>
        {rowStart > 0 ? <div>{rowStart} rows above...</div> : null}
        <div
          style={{
            fontFamily: "monospace",
            overflowX: "scroll",
            display: "flex",
            flexDirection: "row"
          }}
        >
          <span
            style={{
              borderColor: "black",
              borderRight: "1px solid",
              paddingRight: "5px",
              paddingLeft: "5px",
              marginRight: "10px"
            }}
          >
            {data.chunks(HEX_BYTES_PER_ROW).map((row, i) => (
              <div key={i}>
                <span style={{}}>
                  {row.byteStart.toString(16).padStart(8, "0")}
                </span>
              </div>
            ))}
          </span>
          <span>
            {data.chunks(HEX_BYTES_PER_ROW).map((row, i) => (
              <div key={i}>
                {row.chunks(HEX_BYTES_PER_GROUP).map((group, i) => (
                  <span style={{ paddingRight: "10px" }} key={i}>
                    {group.chunks(1).map((byte, i) => {
                      let selected = false;
                      if (this.props.selected) {
                        if (this.props.selected instanceof ByteRange) {
                          selected = this.props.selected.contains(byte);
                        } else if (this.props.selected instanceof BitRange) {
                          selected = this.props.selected
                            .enclosingByteRange()
                            .contains(byte);
                        }
                      }

                      return (
                        <span
                          style={{
                            paddingRight: "5px",
                            backgroundColor: selected ? SELECTED_COLOR : ""
                          }}
                          key={i}
                        >
                          {byte.toHex()}
                        </span>
                      );
                    })}
                  </span>
                ))}
              </div>
            ))}
          </span>
        </div>
        {rowEnd < totalRows ? (
          <div>{totalRows - rowEnd} rows below...</div>
        ) : null}
      </>
    );
  }

  renderBinary() {
    let data = this.props.data;
    if (data.byteLength > this.props.maxRows * BINARY_BYTES_PER_ROW)
      return <div>File is too large for binary view.</div>;

    return (
      <div style={{ fontFamily: "monospace" }}>
        {data.chunks(BINARY_BYTES_PER_ROW).map((row, i) => (
          <div key={i}>
            {row.chunks(BINARY_BYTES_PER_GROUP).map((group, i) => (
              <span style={{ paddingRight: "10px" }} key={i}>
                {group.chunks(1).map((byte, i) => {
                  return (
                    <span
                      style={{
                        paddingRight: "5px"
                      }}
                      key={i}
                    >
                      {byte
                        .bits(0)
                        .chunks(1)
                        .map((b, i) => {
                          let selected = false;
                          if (this.props.selected) {
                            if (this.props.selected instanceof ByteRange) {
                              selected = this.props.selected.contains(
                                b.enclosingByteRange()
                              );
                            } else if (
                              this.props.selected instanceof BitRange
                            ) {
                              selected = this.props.selected.contains(b);
                            }
                          }
                          return (
                            <span
                              style={{
                                backgroundColor: selected ? SELECTED_COLOR : ""
                              }}
                              key={i}
                            >
                              {b.readBool() ? "1" : "0"}
                            </span>
                          );
                        })}
                    </span>
                  );
                })}
              </span>
            ))}
          </div>
        ))}
      </div>
    );
  }
}
