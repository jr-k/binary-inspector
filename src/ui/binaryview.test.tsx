import { BinaryView } from "./binaryview";
import {create} from 'react-test-renderer';
import { ByteRange } from "../core/range";
import * as React from "react";

function makeTestBuffer(size: number): ArrayBuffer {
  const array = new Uint8Array(size);
  for(let i = 0; i < array.length; ++i) {
    array[i] = i;
  }
  return array.buffer;
}

const SHORT_SEQUENCE = makeTestBuffer(32);
const SHORT_SEQUENCE_UNALIGNED = makeTestBuffer(32 - 4);

it('renders a short sequence of hex bytes', () => {
  const data = new ByteRange(SHORT_SEQUENCE, 0, SHORT_SEQUENCE.byteLength);
    const tree = create(<BinaryView data={data}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders a short sequence of unaligned hex bytes', () => {
  const data = new ByteRange(SHORT_SEQUENCE_UNALIGNED, 0, SHORT_SEQUENCE_UNALIGNED.byteLength);
    const tree = create(<BinaryView data={data}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
});
