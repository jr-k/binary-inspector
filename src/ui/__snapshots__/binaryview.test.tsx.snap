// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a long sequence of hex bytes with slicing 1`] = `
Array [
  <form
    className=""
  >
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-2"
      >
        Format
      </label>
      <div
        className="col-sm-10"
      >
        <select
          className="form-control"
          onChange={[Function]}
          value="hex"
        >
          <option
            value="hex"
          >
            Hex
          </option>
          <option
            value="binary"
          >
            Binary
          </option>
        </select>
      </div>
    </div>
  </form>,
  <div
    className="border rounded"
    style={
      Object {
        "maxHeight": "550px",
        "overflowY": "scroll",
      }
    }
  >
    <div>
      1
       rows above...
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "monospace",
          "overflowX": "scroll",
        }
      }
    >
      <span
        style={
          Object {
            "borderColor": "black",
            "borderRight": "1px solid",
            "marginRight": "10px",
            "paddingLeft": "5px",
            "paddingRight": "5px",
          }
        }
      >
        <div
          data-byte-start="16"
        >
          <span
            style={Object {}}
          >
            00000010
          </span>
        </div>
        <div
          data-byte-start="32"
        >
          <span
            style={Object {}}
          >
            00000020
          </span>
        </div>
        <div
          data-byte-start="48"
        >
          <span
            style={Object {}}
          >
            00000030
          </span>
        </div>
      </span>
      <span>
        <div>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              10
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              11
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              12
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              13
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              14
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              15
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              16
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              17
            </span>
          </span>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              18
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              19
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1a
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1b
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1c
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1d
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1e
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1f
            </span>
          </span>
        </div>
        <div>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              20
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              21
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              22
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              23
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              24
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              25
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              26
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              27
            </span>
          </span>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              28
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              29
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              2a
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              2b
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              2c
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              2d
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              2e
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "#dedede",
                  "paddingRight": "5px",
                }
              }
            >
              2f
            </span>
          </span>
        </div>
        <div>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              30
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              31
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              32
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              33
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              34
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              35
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              36
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              37
            </span>
          </span>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              38
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              39
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              3a
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              3b
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              3c
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              3d
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              3e
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              3f
            </span>
          </span>
        </div>
      </span>
    </div>
    <div>
      1
       rows below...
    </div>
  </div>,
]
`;

exports[`renders a short sequence of hex bytes 1`] = `
Array [
  <form
    className=""
  >
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-2"
      >
        Format
      </label>
      <div
        className="col-sm-10"
      >
        <select
          className="form-control"
          onChange={[Function]}
          value="hex"
        >
          <option
            value="hex"
          >
            Hex
          </option>
          <option
            value="binary"
          >
            Binary
          </option>
        </select>
      </div>
    </div>
  </form>,
  <div
    className="border rounded"
    style={
      Object {
        "maxHeight": "550px",
        "overflowY": "scroll",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "monospace",
          "overflowX": "scroll",
        }
      }
    >
      <span
        style={
          Object {
            "borderColor": "black",
            "borderRight": "1px solid",
            "marginRight": "10px",
            "paddingLeft": "5px",
            "paddingRight": "5px",
          }
        }
      >
        <div
          data-byte-start="0"
        >
          <span
            style={Object {}}
          >
            00000000
          </span>
        </div>
        <div
          data-byte-start="16"
        >
          <span
            style={Object {}}
          >
            00000010
          </span>
        </div>
      </span>
      <span>
        <div>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              00
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              01
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              02
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              03
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              04
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              05
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              06
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              07
            </span>
          </span>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              08
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              09
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0a
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0b
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0c
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0d
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0e
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0f
            </span>
          </span>
        </div>
        <div>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              10
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              11
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              12
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              13
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              14
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              15
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              16
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              17
            </span>
          </span>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              18
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              19
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1a
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1b
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1c
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1d
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1e
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1f
            </span>
          </span>
        </div>
      </span>
    </div>
  </div>,
]
`;

exports[`renders a short sequence of unaligned hex bytes 1`] = `
Array [
  <form
    className=""
  >
    <div
      className="form-group row"
    >
      <label
        className="form-label col-form-label col-sm-2"
      >
        Format
      </label>
      <div
        className="col-sm-10"
      >
        <select
          className="form-control"
          onChange={[Function]}
          value="hex"
        >
          <option
            value="hex"
          >
            Hex
          </option>
          <option
            value="binary"
          >
            Binary
          </option>
        </select>
      </div>
    </div>
  </form>,
  <div
    className="border rounded"
    style={
      Object {
        "maxHeight": "550px",
        "overflowY": "scroll",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "monospace",
          "overflowX": "scroll",
        }
      }
    >
      <span
        style={
          Object {
            "borderColor": "black",
            "borderRight": "1px solid",
            "marginRight": "10px",
            "paddingLeft": "5px",
            "paddingRight": "5px",
          }
        }
      >
        <div
          data-byte-start="0"
        >
          <span
            style={Object {}}
          >
            00000000
          </span>
        </div>
        <div
          data-byte-start="16"
        >
          <span
            style={Object {}}
          >
            00000010
          </span>
        </div>
      </span>
      <span>
        <div>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              00
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              01
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              02
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              03
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              04
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              05
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              06
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              07
            </span>
          </span>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              08
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              09
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0a
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0b
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0c
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0d
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0e
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              0f
            </span>
          </span>
        </div>
        <div>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              10
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              11
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              12
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              13
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              14
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              15
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              16
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              17
            </span>
          </span>
          <span
            style={
              Object {
                "paddingRight": "10px",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              18
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              19
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1a
            </span>
            <span
              style={
                Object {
                  "backgroundColor": "",
                  "paddingRight": "5px",
                }
              }
            >
              1b
            </span>
          </span>
        </div>
      </span>
    </div>
  </div>,
]
`;
